notification-hubs-sdk/src/main/java/com/microsoft/windowsazure/messaging/notificationhubs/DebounceInstallationAdapter.java

```diff
@@ -0,0 +1,61 @@
+package com.microsoft.windowsazure.messaging.notificationhubs;
+
+import android.content.Context;
+import android.content.SharedPreferences;
+
+import com.microsoft.windowsazure.messaging.R;
+
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
+
+public class DebounceInstallationAdapter implements InstallationAdapter {
+
+    private static final String PREFERENCE_KEY = "recentInstallation";
+    private final ScheduledExecutorService mScheduler = Executors.newScheduledThreadPool(1);
+    private InstallationAdapter mInstallationAdapter;
+    private long mInterval;
+    private ScheduledFuture<?> mSchedFuture;
+    private SharedPreferences mPreferences;
+
+    public DebounceInstallationAdapter(InstallationAdapter installationAdapter) {
+        this(installationAdapter, 2000);
+    }
+
+    public DebounceInstallationAdapter(InstallationAdapter installationAdapter, long interval) {
+        super();
+        this.mInstallationAdapter = installationAdapter;
+        mInterval = interval;
+    }
+
+    private void setPreferences(Context context) {
+        mPreferences = context.getSharedPreferences(context.getString(R.string.installation_enrichment_file_key), Context.MODE_MULTI_PROCESS);
+    }
+
+    @Override
+    public void saveInstallation(Context context, Installation installation) {
+        if (mPreferences == null) {
+            setPreferences(context);
+        }
+
+        if (mSchedFuture != null && !mSchedFuture.isDone()) {
+            mSchedFuture.cancel(true);
+        }
+
+        int recentHash = mPreferences.getInt(PREFERENCE_KEY, 0);
+        if (recentHash != 0 && recentHash == installation.hashCode()) {
```

Comment:
We should be careful comparing the hash to `0`. Can we prove that `0` isn't a valid hash value?

Reply:
Just using the hash really isn't enough. The hash is an `int`, so there will be unpredictable collisions. I think we need to revisit this.

Reply:
Good call. How about we come up with a string format that incorporates all aspects of an installation and get a SHA1 hash? I've used [this approach to great effect](https://github.com/marstr/envelopes/blob/ca84d907a0f922268d8e93b3139fcb819fc30fa8/transaction.go#L38-L44) in one of my personal Go projects, I think it could work well here too. :)

Reply:
I think that we'll see few enough collisions for this to be the approach in preview1. However, I've also added a bug card to our internal tracker to re-evaluate this.